File: .github/scripts/generate_metadata.py
import hashlib
import logging
import os
import shutil
import yaml
from anthropic import Anthropic

# Set up logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Path to the analyzer prompt file
ANALYZER_PROMPT_PATH = '.github/prompts/ai_prompt_analyzer_and_output_generator/prompt.md'

def load_analyzer_prompt():
    """Load the content of the analyzer prompt file."""
    logger.info(f"Loading analyzer prompt from {ANALYZER_PROMPT_PATH}")
    with open(ANALYZER_PROMPT_PATH, 'r') as f:
        content = f.read()
    logger.info(f"Analyzer prompt loaded, length: {len(content)} characters")
    return content

def generate_metadata(prompt_content):
    """Generate metadata for a given prompt content using the Anthropic API."""
    logger.info("Starting metadata generation")
    
    # Check for the presence of the API key
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        logger.error("ANTHROPIC_API_KEY is not set in the environment.")
        raise EnvironmentError("ANTHROPIC_API_KEY is not set in the environment.")
    else:
        logger.info("ANTHROPIC_API_KEY is set.")
    
    # Initialize the Anthropic client
    client = Anthropic(api_key=api_key)
    logger.info("Anthropic client initialized")
    
    # Load the analyzer prompt
    analyzer_prompt = load_analyzer_prompt()
    
    # Create a message using the Anthropic API
    logger.info("Sending request to Anthropic API")
    message = client.messages.create(
        model="claude-3-5-sonnet-20240620",
        max_tokens=2500,
        messages=[
            {
                "role": "user",
                "content": analyzer_prompt.replace("{{PROMPT}}", prompt_content)
            }
        ]
    )
    logger.info("Received response from Anthropic API")
    
    # Log the structure of the response
    logger.info(f"Response structure: {type(message)}")
    logger.info(f"Content structure: {type(message.content)}")
    
    # Extract the YAML content from the AI response
    content = message.content[0].text if isinstance(message.content, list) else message.content
    logger.info(f"Extracted content: {content[:100]}...") # Log first 100 characters
    
    output_start = content.find("<output>")
    output_end = content.find("</output>")
    if output_start != -1 and output_end != -1:
        yaml_content = content[output_start + 8:output_end].strip()
        logger.info(f"Extracted YAML content: {yaml_content[:100]}...") # Log first 100 characters
        metadata = yaml.safe_load(yaml_content)
        logger.info("YAML content parsed successfully")
    else:
        logger.error("Could not find metadata output in AI response")
        raise ValueError("Could not find metadata output in AI response")
    
    logger.info("Metadata generation completed successfully")
    return metadata

def should_update_metadata(prompt_file, metadata_file):
    """Check if metadata should be updated based on content hash or force flag."""
    force_regenerate = os.environ.get('FORCE_REGENERATE', 'false').lower() == 'true'
    
    if force_regenerate:
        logger.info("Forcing metadata regeneration due to system prompt changes.")
        return True, None

    # Generate hash of the prompt file content
    with open(prompt_file, 'rb') as f:
        prompt_content = f.read()
    prompt_hash = hashlib.md5(prompt_content).hexdigest()

    # If metadata file doesn't exist, update is needed
    if not os.path.exists(metadata_file):
        logger.info(f"Metadata file {metadata_file} does not exist. Update needed.")
        return True, prompt_hash

    # Read the stored hash from metadata file
    with open(metadata_file, 'r') as f:
        metadata_content = f.read()
    
    # Extract the stored hash
    stored_hash_line = next((line for line in metadata_content.split('\n') if line.startswith('content_hash:')), None)
    
    if stored_hash_line:
        stored_hash = stored_hash_line.split(':', 1)[1].strip()
        
        # Compare the hashes
        if prompt_hash != stored_hash:
            logger.info(f"Content hash mismatch for {prompt_file}. Update needed.")
            return True, prompt_hash
    else:
        # If no hash found in metadata, update is needed
        logger.info(f"No content hash found in {metadata_file}. Update needed.")
        return True, prompt_hash

    logger.info(f"Content hash match for {prompt_file}. No update needed.")
    return False, prompt_hash

def update_metadata_hash(metadata_file, new_hash):
    """Update or append the content hash in the metadata file."""
    if os.path.exists(metadata_file):
        with open(metadata_file, 'r') as f:
            lines = f.readlines()
        
        # Update or add the hash line
        hash_updated = False
        for i, line in enumerate(lines):
            if line.startswith('content_hash:'):
                lines[i] = f'content_hash: {new_hash}\n'
                hash_updated = True
                break
        
        if not hash_updated:
            lines.append(f'content_hash: {new_hash}\n')
        
        with open(metadata_file, 'w') as f:
            f.writelines(lines)
        logger.info(f"Content hash updated in {metadata_file}")
    else:
        # If metadata file doesn't exist, create it with the hash
        with open(metadata_file, 'w') as f:
            f.write(f'content_hash: {new_hash}\n')
        logger.info(f"New metadata file created with content hash: {metadata_file}")

def update_prompt_metadata():
    """Update metadata for all prompts in the 'prompts' directory."""
    logger.info("Starting update_prompt_metadata process")
    prompts_dir = 'prompts'
    
    # Process the main prompt.md file in the prompts directory
    main_prompt_file = os.path.join(prompts_dir, 'prompt.md')
    if os.path.exists(main_prompt_file):
        logger.info("Processing main prompt.md file")
        with open(main_prompt_file, 'rb') as f:
            prompt_content = f.read()
        metadata = generate_metadata(prompt_content.decode('utf-8'))
        new_dir_name = metadata['directory']
        new_dir_path = os.path.join(prompts_dir, new_dir_name)
        
        # Create new directory and move the prompt file
        logger.info(f"Creating new directory: {new_dir_path}")
        os.makedirs(new_dir_path, exist_ok=True)
        new_prompt_file = os.path.join(new_dir_path, 'prompt.md')
        logger.info(f"Moving {main_prompt_file} to {new_prompt_file}")
        shutil.move(main_prompt_file, new_prompt_file)
        
        # Save metadata
        metadata_path = os.path.join(new_dir_path, 'metadata.yml')
        logger.info(f"Saving metadata to {metadata_path}")
        with open(metadata_path, 'w') as f:
            yaml.dump(metadata, f, sort_keys=False)
        
        # Update content hash
        new_hash = hashlib.md5(prompt_content).hexdigest()
        update_metadata_hash(metadata_path, new_hash)
    
    # Process subdirectories
    for item in os.listdir(prompts_dir):
        item_path = os.path.join(prompts_dir, item)
        
        if os.path.isdir(item_path):
            logger.info(f"Processing directory: {item}")
            prompt_file = os.path.join(item_path, 'prompt.md')
            metadata_file = os.path.join(item_path, 'metadata.yml')
            
            if os.path.exists(prompt_file):
                should_update, new_hash = should_update_metadata(prompt_file, metadata_file)
                if should_update:
                    logger.info(f"Updating metadata for {item}")
                    with open(prompt_file, 'r') as f:
                        prompt_content = f.read()
                    
                    metadata = generate_metadata(prompt_content)
                    new_dir_name = metadata['directory']
                    
                    # Rename directory if necessary
                    if new_dir_name != item:
                        new_dir_path = os.path.join(prompts_dir, new_dir_name)
                        logger.info(f"Renaming directory from {item} to {new_dir_name}")
                        if os.path.exists(new_dir_path):
                            logger.warning(f"Directory {new_dir_name} already exists. Updating contents.")
                            for file in os.listdir(item_path):
                                src = os.path.join(item_path, file)
                                dst = os.path.join(new_dir_path, file)
                                if os.path.isfile(src):
                                    shutil.copy2(src, dst)
                            shutil.rmtree(item_path)
                        else:
                            os.rename(item_path, new_dir_path)
                        item_path = new_dir_path
                    
                    # Save updated metadata
                    metadata_path = os.path.join(item_path, 'metadata.yml')
                    logger.info(f"Saving updated metadata to {metadata_path}")
                    with open(metadata_path, 'w') as f:
                        yaml.dump(metadata, f, sort_keys=False)
                    
                    # Update content hash
                    update_metadata_hash(metadata_path, new_hash)
                else:
                    logger.info(f"Metadata for {item} is up to date")
            else:
                logger.warning(f"No prompt.md file found in {item_path}")
    
    logger.info("update_prompt_metadata process completed")

if __name__ == '__main__':
    logger.info("Script started")
    try:
        update_prompt_metadata()
        logger.info("Script completed successfully")
    except Exception as e:
        logger.exception(f"An error occurred: {str(e)}")
        raise

File: .github/scripts/generate_metadata.ts
import {
    copyFile,
    createDirectory,
    fileExists,
    isDirectory,
    isFile,
    readDirectory,
    readFileContent,
    removeDirectory,
    renameFile,
    writeFileContent,
} from './utils/file_operations';
import { initializeAnthropicClient, sendAnthropicRequest } from './utils/anthropic_client';
import { parseYamlContent, dumpYamlContent, sanitizeYamlContent } from './utils/yaml_operations';
import { Metadata } from './types/metadata';
import * as path from 'path';
import * as crypto from 'crypto';
import { ContentBlock, Message } from '@anthropic-ai/sdk/resources';
import config from './config';
import logger from './utils/logger';
import Anthropic from '@anthropic-ai/sdk';

/**
 * Loads the analyzer prompt from the specified file.
 * @returns The content of the analyzer prompt.
 */
async function loadAnalyzerPrompt(): Promise<string> {
    logger.info(`Loading analyzer prompt from ${config.ANALYZER_PROMPT_PATH}`);
    const content = await readFileContent(config.ANALYZER_PROMPT_PATH);
    logger.info(`Analyzer prompt loaded, length: ${content.length} characters`);
    return content;
}

/**
 * Extracts the content from an Anthropic API message.
 * @param message - The message object from the Anthropic API.
 * @returns The extracted content as a string.
 */
function extractContentFromMessage(message: Message): string {
    if (!message.content || message.content.length === 0) {
        return '';
    }

    return message.content
        .map((block: ContentBlock) => {
            if (block.type === 'text') {
                return block.text;
            } else if (block.type === 'tool_use') {
                return `[Tool Use: ${block.name}]\nInput: ${JSON.stringify(block.input)}`;
            }
            return JSON.stringify(block);
        })
        .join('\n');
}

/**
 * Extracts YAML content from a string between <output> tags.
 * @param content - The string containing YAML content.
 * @returns The extracted YAML content.
 */
function extractYamlContent(content: string): string {
    const outputStart = content.indexOf('<output>');
    const outputEnd = content.indexOf('</output>');
    return outputStart !== -1 && outputEnd !== -1 ? content.slice(outputStart + 8, outputEnd).trim() : content.trim();
}

/**
 * Processes the prompt content using the Anthropic API and generates metadata.
 * @param client - The Anthropic API client.
 * @param analyzerPrompt - The analyzer prompt.
 * @param promptContent - The content of the prompt to analyze.
 * @returns The generated metadata.
 */
async function processPromptContent(
    client: Anthropic,
    analyzerPrompt: string,
    promptContent: string,
): Promise<Metadata> {
    const message = await sendAnthropicRequest(client, analyzerPrompt.replace('{{PROMPT}}', promptContent));
    const content = extractContentFromMessage(message);
    const yamlContent = extractYamlContent(content);
    return parseYamlContent(yamlContent);
}

/**
 * Generates metadata for a given prompt content.
 * @param promptContent - The content of the prompt.
 * @returns The generated metadata.
 */
export async function generateMetadata(promptContent: string): Promise<Metadata> {
    logger.info('Starting metadata generation');

    const client = initializeAnthropicClient();
    logger.info('Anthropic client initialized');

    try {
        const analyzerPrompt = await loadAnalyzerPrompt();
        return await processPromptContent(client, analyzerPrompt, promptContent);
    } catch (error) {
        logger.error('Error in generateMetadata:', error);
        throw error;
    }
}

/**
 * Determines if metadata should be updated based on content hash.
 * @param promptFile - Path to the prompt file.
 * @param metadataFile - Path to the metadata file.
 * @returns A tuple indicating if update is needed and the new hash.
 */
export async function shouldUpdateMetadata(promptFile: string, metadataFile: string): Promise<[boolean, string]> {
    const forceRegenerate = process.env[config.FORCE_REGENERATE_ENV_VAR]?.toLowerCase() === 'true';

    const promptContent = await readFileContent(promptFile);
    const promptHash = crypto.createHash('md5').update(promptContent).digest('hex');

    if (forceRegenerate) {
        logger.info('Forcing metadata regeneration due to system prompt changes.');
        return [true, promptHash];
    }

    try {
        const metadataContent = await readFileContent(metadataFile);
        const storedHashLine = metadataContent.split('\n').find((line) => line.trim().startsWith('content_hash:'));

        if (storedHashLine) {
            const storedHash = storedHashLine.split(':')[1].trim();
            if (promptHash !== storedHash) {
                logger.info(`Content hash mismatch for ${promptFile}. Update needed.`);
                return [true, promptHash];
            }
        } else {
            logger.info(`No content hash found in ${metadataFile}. Update needed.`);
            return [true, promptHash];
        }

        logger.info(`Content hash match for ${promptFile}. No update needed.`);
        return [false, promptHash];
    } catch (error) {
        if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
            logger.info(`Metadata file ${metadataFile} does not exist. Update needed.`);
            return [true, promptHash];
        }
        throw error;
    }
}

/**
 * Updates the content hash in the metadata file.
 * @param metadataFile - Path to the metadata file.
 * @param newHash - The new hash to be updated.
 */
export async function updateMetadataHash(metadataFile: string, newHash: string): Promise<void> {
    try {
        const content = await readFileContent(metadataFile);
        const lines = content.split('\n');

        let hashUpdated = false;
        for (let i = 0; i < lines.length; i++) {
            if (lines[i].trim().startsWith('content_hash:')) {
                lines[i] = `content_hash: ${newHash}`;
                hashUpdated = true;
                break;
            }
        }
        if (!hashUpdated) {
            lines.push(`content_hash: ${newHash}`);
        }
        await writeFileContent(metadataFile, sanitizeYamlContent(lines.join('\n')));
        logger.info(`Content hash updated in ${metadataFile}`);
    } catch (error) {
        logger.error(`Failed to update content hash for ${metadataFile}`, error);
        throw error;
    }
}

/**
 * Updates metadata for all prompts in the library.
 */
export async function updatePromptMetadata(): Promise<void> {
    logger.info('Starting update_prompt_metadata process');

    await processMainPrompt(config.PROMPTS_DIR);
    await processPromptDirectories(config.PROMPTS_DIR);

    logger.info('update_prompt_metadata process completed');
}

/**
 * Processes the main prompt file if it exists.
 * @param promptsDir - The directory containing prompts.
 */
async function processMainPrompt(promptsDir: string): Promise<void> {
    const mainPromptFile = path.join(promptsDir, config.PROMPT_FILE_NAME);
    if (await fileExists(mainPromptFile)) {
        logger.info('Processing main prompt.md file');
        const promptContent = await readFileContent(mainPromptFile);
        const metadata = await generateMetadata(promptContent);
        const newDirName = metadata.directory;
        const newDirPath = path.join(promptsDir, newDirName);

        await createDirectory(newDirPath);
        const newPromptFile = path.join(newDirPath, config.PROMPT_FILE_NAME);
        await renameFile(mainPromptFile, newPromptFile);

        const metadataPath = path.join(newDirPath, config.METADATA_FILE_NAME);
        await writeFileContent(metadataPath, sanitizeYamlContent(dumpYamlContent(metadata)));

        const newHash = crypto.createHash('md5').update(promptContent).digest('hex');
        await updateMetadataHash(metadataPath, newHash);
    }
}

/**
 * Processes all prompt directories in the library.
 * @param promptsDir - The directory containing prompts.
 */
async function processPromptDirectories(promptsDir: string): Promise<void> {
    const items = await readDirectory(promptsDir);
    await Promise.all(
        items.map(async (item) => {
            const currentItemPath = path.join(promptsDir, item);

            if (await isDirectory(currentItemPath)) {
                logger.info(`Processing directory: ${item}`);
                const promptFile = path.join(currentItemPath, config.PROMPT_FILE_NAME);
                const metadataFile = path.join(currentItemPath, config.METADATA_FILE_NAME);

                if (await fileExists(promptFile)) {
                    await processPromptFile(promptFile, metadataFile, currentItemPath, promptsDir, item);
                } else {
                    logger.warn(`No ${config.PROMPT_FILE_NAME} file found in ${currentItemPath}`);
                }
            }
        }),
    );
}

/**
 * Processes a single prompt file and updates its metadata if necessary.
 * @param promptFile - Path to the prompt file.
 * @param metadataFile - Path to the metadata file.
 * @param currentItemPath - Current directory path of the prompt.
 * @param promptsDir - The directory containing all prompts.
 * @param item - The name of the current prompt directory.
 */
async function processPromptFile(
    promptFile: string,
    metadataFile: string,
    currentItemPath: string,
    promptsDir: string,
    item: string,
): Promise<void> {
    const [shouldUpdate, newHash] = await shouldUpdateMetadata(promptFile, metadataFile);
    if (shouldUpdate) {
        logger.info(`Updating metadata for ${item}`);
        const promptContent = await readFileContent(promptFile);

        try {
            const metadata = await generateMetadata(promptContent);
            if (!metadata || Object.keys(metadata).length === 0) {
                logger.warn(`Failed to generate metadata for ${item}. Skipping.`);
                return;
            }
            const newDirName = metadata.directory;

            if (newDirName !== item) {
                await updatePromptDirectory(promptsDir, item, newDirName, currentItemPath);
                currentItemPath = path.join(promptsDir, newDirName);
            }

            const metadataPath = path.join(currentItemPath, config.METADATA_FILE_NAME);
            await writeFileContent(metadataPath, sanitizeYamlContent(dumpYamlContent(metadata)));

            await updateMetadataHash(
                metadataPath,
                newHash || crypto.createHash('md5').update(promptContent).digest('hex'),
            );
        } catch (error) {
            logger.error(`Error processing ${item}:`, error);
        }
    } else {
        logger.info(`Metadata for ${item} is up to date`);
    }
}

/**
 * Updates the prompt directory name if it has changed.
 * @param promptsDir - The directory containing all prompts.
 * @param oldDirName - The current directory name.
 * @param newDirName - The new directory name.
 * @param currentItemPath - The current path of the prompt directory.
 */
async function updatePromptDirectory(
    promptsDir: string,
    oldDirName: string,
    newDirName: string,
    currentItemPath: string,
): Promise<void> {
    const newDirPath = path.join(promptsDir, newDirName);
    logger.info(`Renaming directory from ${oldDirName} to ${newDirName}`);
    if (await fileExists(newDirPath)) {
        logger.warn(`Directory ${newDirName} already exists. Updating contents.`);
        const files = await readDirectory(currentItemPath);
        await Promise.all(
            files.map(async (file) => {
                const src = path.join(currentItemPath, file);
                const dst = path.join(newDirPath, file);
                if (await isFile(src)) {
                    await copyFile(src, dst);
                }
            }),
        );
        await removeDirectory(currentItemPath);
    } else {
        await renameFile(currentItemPath, newDirPath);
    }
}

if (require.main === module) {
    updatePromptMetadata().catch((error) => {
        logger.error('Error in main execution:', error);
        process.exit(1);
    });
}
